ql/pricingengines/vanilla/qdfpamericanengine.hpp

```diff
@@ -0,0 +1,151 @@
+         n: number of Chebyshev nodes to interpolate the exercise boundary
+         p: order of Gauss-Legendre integration in final conversion of the
+            exercise boundary into option prices
+    */
+    class QdFpLegendreScheme: public QdFpIterationScheme {
+      public:
+        QdFpLegendreScheme(Size l, Size m, Size n, Size p);
+
+        Size getNumberOfChebyshevInterpolationNodes() const override;
+        Size getNumberOfNaiveFixedPointSteps() const override;
+        Size getNumberOfJacobiNewtonFixedPointSteps() const override;
+
+        ext::shared_ptr<Integrator> getFixedPointIntegrator() const override;
+        ext::shared_ptr<Integrator> getExerciseBoundaryToPriceIntegrator() const override;
+
+      private:
+        const Size m_, n_;
+        const ext::shared_ptr<Integrator> fpIntegrator_;
+        const ext::shared_ptr<Integrator> exerciseBoundaryIntegrator_;
+    };
+
+    //! Legendre-Tanh-Sinh (l,m,n)-eps Scheme
+    /*!  l: order of Gauss-Legendre integration within every fixed point iteration
+            step.
+         m: fixed point iteration steps, first step is a partial Jacobi-Newton,
+            the rest are naive Richardson fixed point iterations
+         n: number of Chebyshev nodes to interpolate the exercise boundary
+         eps: final conversion of the exercise boundary into option prices
+               is carried out by a tanh-sinh integration with accuracy eps
+    */
+    class QdFpLegendreTanhSinhScheme: public QdFpLegendreScheme {
+      public:
+        QdFpLegendreTanhSinhScheme(Size l, Size m, Size n, Real eps);
+
+        ext::shared_ptr<Integrator> getExerciseBoundaryToPriceIntegrator() const override;
+
+      private:
+        const Real eps_;
+    };
+
+    //! tanh-sinh (m,n)-eps Scheme
+    /*!  m: fixed point iteration steps, first step is a partial Jacobi-Newton,
+            the rest are naive Richardson fixed point iterations
+         n: number of Chebyshev nodes to interpolate the exercise boundary
+         eps: tanh-sinh integration precision
+    */
+    class QdFpTanhSinhIterationScheme: public QdFpIterationScheme {
+      public:
+        QdFpTanhSinhIterationScheme(Size m, Size n, Real eps);
+
+        Size getNumberOfChebyshevInterpolationNodes() const override;
+        Size getNumberOfNaiveFixedPointSteps() const override;
+        Size getNumberOfJacobiNewtonFixedPointSteps() const override;
+
+        ext::shared_ptr<Integrator> getFixedPointIntegrator() const override;
+        ext::shared_ptr<Integrator> getExerciseBoundaryToPriceIntegrator() const override;
+      private:
+        const Size m_, n_;
+        const ext::shared_ptr<Integrator> integrator_;
+    };
+
+    class QdFpIterationSchemeStdFactory {
+      public:
+        static ext::shared_ptr<QdFpIterationScheme> fastScheme();
+        static ext::shared_ptr<QdFpIterationScheme> accurateScheme();
+        static ext::shared_ptr<QdFpIterationScheme> highPrecisionScheme();
```

Comment:
Any objections if I move these three methods into `QdFpIterationScheme` and avoid a separate factory class?

Reply:
Or possibly in the engine itself?

Reply:
sure, the engine itself looks like a better place for these functions.